miden-lib/asm/miden/note.masm

```diff
@@ -143,4 +143,435 @@ export.get_serial_number
+
+    # hash remaining input values if there any left
+    # if num_inputs%8 is not ZERO
+    dup.14 push.0 neq
+    if.true
+        # if there are exactly 4 elements left to hash, just pad them with [0, 0, 0, 1] Word
+        dup.14 eq.4
+        if.true
+            # load the remaining word
+            dropw movup.8 mem_loadw
+            # => [B'', A', end_addr, num_inputs%8]
+
+            # clean the stack
+            movupw.2 dropw
+            # => [B'', A', end_addr, num_inputs%8]
+
+            # push the capacity word
+            push.1.0.0.0
+            hperm
+            # => [F, E, D]
+
+            exec.native::state_to_digest
+            # => [E]
+        else
+            # load the remaining double word
+            mem_stream
+            # => [E, D, A', inputs_ptr'+2, end_addr, num_inputs%8]
+
+            # clean the stack
+            movup.12 drop movup.12 drop
+            # => [E, D, A', num_inputs%8]
+
+            # get the number of elements we need to drop
+            push.8 movup.13 u32wrapping_sub movdn.12
+            # => [E, D, A', drop_counter]
+
+            # get the number of zeros we will need to pad
+            dup.12 u32wrapping_sub.1 movdn.13
+            # => [E, D, A', drop_counter, zero_counter]
+
+            # check the drop_counter
+            dup.12 neq.0
+            while.true
+                # decrease the counter
+                movup.12 u32wrapping_sub.1 movdn.12
+
+                # drop the top element, push 0 to the stack and move it down to maintain the overall
+                # number of elements on stack to be able to access the drop_counter at position 12
+                drop push.0 movdn.11
+
+                # check the drop_counter
+                dup.12 neq.0
+            end
+            # => [a, ..., x, A', 0, ..., 0, 0, zero_counter]
+            #                               ^-- 12th position
+
+            # push the dividing 1 (move the previously added zero and incr it)
+            movup.12 add.1
+            # => [1, a, ..., x, A', 0, ..., 0, 0, zero_counter]
+            #                                  ^-- 12th position
+
+            # check the zero_counter
+            dup.13 neq.0
+            while.true
+                # decrease the counter
+                movup.13 u32wrapping_sub.1 movdn.13
+
+                # move the previously added zero
+                movup.12
+
+                # check the zero_counter
+                dup.13 neq.0
+            end
+            # => [C, B, A', 0, 0, ... ]
+            
+            hperm
+            # => [F, E, D]
+
+            exec.native::state_to_digest
+            # => [E]
+        end
+    else
+        dropw movdnw.2 dropw dropw
+        # => [B']
+    end
+end
+
+#! Computes hash of note inputs starting at the specified memory address.
+#!
+#! Inputs:  [inputs_ptr, num_inputs]
+#! Outputs: [HASH]
+#!
+#! Panics if num_inputs is greater than 128.
+export.compute_inputs_hash_cdrop
```

Comment:
Let's go with this version. It will get somewhat simpler when we change the padding rule.

Also, we should document what is returned if the number of inputs is 0 (it should be `EMPTY_WORD`, right?).

Lastly, let's add a `TODO` here to mention that most of this function will need to be replaced with a `hash_memory` function from `miden-base` when we migrate to the next version of `miden-vm`.